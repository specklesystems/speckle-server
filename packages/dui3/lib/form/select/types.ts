export type ProjectsSelectItemType = { id?: string; name: string }
export type ModelsSelectItemType = { id?: string; name: string }
export type VersionsSelectItemType = {
  id: string
  message: string
  createdAt: Date
  referencedObject: string
  sourceApplication: string
}
