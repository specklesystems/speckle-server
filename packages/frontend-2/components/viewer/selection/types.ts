export type KeyValuePair = {
  key: string
  value: unknown
  units?: string
  type?: string
  innerType?: string | null
  arrayLength?: number | null
  arrayPreview?: string | null
  backendPath?: string
}
