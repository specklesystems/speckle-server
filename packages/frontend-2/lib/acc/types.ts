import type { AccHub, AccItem } from '@speckle/shared/acc'

// TODO ACC: Replace with type information inferred from gql queries, if possible
export type AccSyncItem = {
  id: string
  accHub: AccHub
  accHubId: string
  createdBy: string
  projectId: string
  modelId: string
  projectName: string
  modelName: string
  accItem: AccItem
  status: AccSyncItemStatus
}

export type AccSyncItemStatus =
  | 'pending'
  | 'syncing'
  | 'paused'
  | 'failed'
  | 'succeeded'
