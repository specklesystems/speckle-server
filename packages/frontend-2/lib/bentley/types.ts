import type { Integration } from '~/lib/integrations/types'
import bentleyLogo from '~/assets/images/integrations/bentley.png'

export const BentleyIntegration: Integration = {
  cookieKey: 'bentley_itwin_tokens',
  name: 'Bentley ProjectWise',
  description: 'Sync your files in Bentley ProjectWise into Speckle.',
  logo: bentleyLogo,
  connected: false,
  enabled: true, // TODO: do false later when it is a
  status: 'notConnected'
}
