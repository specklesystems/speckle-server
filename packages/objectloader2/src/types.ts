export type CustomLogger = (message?: string, ...optionalParams: unknown[]) => void

export interface Item {
  id: string
  obj: Base
}

export interface Base {
  id: string
  __closure?: string[]
}

export function isString(value: unknown): value is string {
  return typeof value === 'string'
}
