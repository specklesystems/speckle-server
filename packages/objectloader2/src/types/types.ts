export type CustomLogger = (message?: string, ...optionalParams: unknown[]) => void

export type Fetcher = (
  input: string | URL | Request,
  init?: RequestInit
) => Promise<Response>

export interface Item {
  baseId: string
  base: Base
  size?: number
}

export interface Base {
  id: string
  speckle_type: string
  __closure?: Record<string, number>
}

export interface Reference {
  speckle_type: string
  referencedId: string
  __closure?: Record<string, number>
}

export interface DataChunk extends Base {
  data?: Base[]
}

export function isBase(maybeBase?: unknown): maybeBase is Base {
  return (
    maybeBase !== null &&
    typeof maybeBase === 'object' &&
    'id' in maybeBase &&
    typeof maybeBase.id === 'string'
  )
}

export function isReference(maybeRef?: unknown): maybeRef is Reference {
  return (
    maybeRef !== null &&
    typeof maybeRef === 'object' &&
    'referencedId' in maybeRef &&
    typeof maybeRef.referencedId === 'string'
  )
}

export function isScalar(
  value: unknown
): value is string | number | boolean | bigint | symbol | undefined {
  const type = typeof value
  return (
    value === null ||
    type === 'string' ||
    type === 'number' ||
    type === 'boolean' ||
    type === 'bigint' ||
    type === 'symbol' ||
    type === 'undefined'
  )
}

export function take<T>(it: Iterator<T>, count: number): T[] {
  const result: T[] = []
  for (let i = 0; i < count; i++) {
    const itr = it.next()
    if (itr.done) break
    result.push(itr.value)
  }
  return result
}
