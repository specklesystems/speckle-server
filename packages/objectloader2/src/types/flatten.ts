import { Base } from './types.js'

/**
 * Flattens a speckle object. It will ignore arrays, null and undefined valuesm, as well as various 'safe to ignore' speckle properties, such as
 * bbox, __closure, __parents, totalChildrenCount.
 * @param obj object to flatten
 * @returns an object with all its props flattened into `prop.subprop.subsubprop`.
 */
export function flattenBase(obj: Base): Record<string, string | number> {
  const flatten = {} as Record<string, string | number>
  for (const k in obj) {
    if (
      k === 'id' ||
      k === '__closure' ||
      k === '__parents' ||
      k === 'bbox' ||
      k === 'totalChildrenCount'
    ) {
      continue
    }

    const v = obj[k]
    if (v === null || v === undefined || Array.isArray(v)) {
      continue
    }
    if (v.constructor === Object) {
      const flattenProp = flattenBase(v as unknown as Base)
      for (const pk in flattenProp) {
        flatten[k + '.' + pk] = flattenProp[pk]
      }
      continue
    }
    const type = typeof v
    if (type === 'string' || type === 'number') {
      flatten[k] = v as string | number
    } else if (type === 'boolean') {
      flatten[k] = v ? 'true' : 'false' // Convert boolean to string for consistency
    }
  }
  if (obj.id) {
    flatten.id = obj.id
  }
  return flatten
}

export function sententizeBase(obj: Base): string {
  const flattened = flattenBase(obj)
  const propertyStrings = Object.entries(flattened).map(([key, value]) => {
    // Make the key more readable (e.g., "Fire Rating" instead of "FireRating")
    const formattedKey = key.replace(/([A-Z])/g, ' $1').trim()
    return `${formattedKey} is ${value}`
  })
  // Join the individual strings with a comma and a space.
  return `This object has the following properties: ${propertyStrings.join(', ')}.`
}
