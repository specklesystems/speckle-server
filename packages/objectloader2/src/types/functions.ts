import { Base, Reference } from './types.js'

export type CustomLogger = (message?: string, ...optionalParams: unknown[]) => void

export type Fetcher = (
  input: string | URL | Request,
  init?: RequestInit
) => Promise<Response>

export function isBase(maybeBase?: unknown): maybeBase is Base {
  return (
    maybeBase !== null &&
    typeof maybeBase === 'object' &&
    'id' in maybeBase &&
    typeof maybeBase.id === 'string'
  )
}

export function isReference(maybeRef?: unknown): maybeRef is Reference {
  return (
    maybeRef !== null &&
    typeof maybeRef === 'object' &&
    'referencedId' in maybeRef &&
    typeof maybeRef.referencedId === 'string'
  )
}

export function isScalar(
  value: unknown
): value is string | number | boolean | bigint | symbol | undefined {
  const type = typeof value
  return (
    value === null ||
    type === 'string' ||
    type === 'number' ||
    type === 'boolean' ||
    type === 'bigint' ||
    type === 'symbol' ||
    type === 'undefined'
  )
}

export function take<T>(it: Iterator<T>, count: number): T[] {
  const result: T[] = []
  for (let i = 0; i < count; i++) {
    const itr = it.next()
    if (itr.done) break
    result.push(itr.value)
  }
  return result
}

export enum ObjectLoader2Flags {
  DEBUG = 'debug',
  USE_CACHE = 'useCache'
}

const defaultValues: Record<ObjectLoader2Flags, string> = {
  [ObjectLoader2Flags.DEBUG]: 'false',
  [ObjectLoader2Flags.USE_CACHE]: 'true'
}

export function getFeatureFlag(
  paramName: ObjectLoader2Flags,
  useDefault: boolean = true
): string | undefined {
  // Check if the code is running in a browser environment üåê
  const isBrowser =
    typeof window !== 'undefined' && typeof window.document !== 'undefined'

  if (!isBrowser) {
    // If in Node.js or another server environment, return the default
    return useDefault ? defaultValues[paramName] : undefined
  }

  // In a browser, parse the query string
  const params = new URLSearchParams(window.location.search)

  // .get() returns the value, or null if it's not found.
  // The nullish coalescing operator (??) provides the default value
  // if the left-hand side is null or undefined.
  return params.get(paramName) ?? (useDefault ? defaultValues[paramName] : undefined)
}

/**
 * Finds the first index of a "needle" Uint8Array within a "haystack" Uint8Array.
 * @param haystack The larger array to search within.
 * @param needle The smaller array to search for.
 * @param start The index to start searching from. Defaults to 0.
 * @returns The starting index of the needle, or -1 if not found.
 */
export function indexOf(
  haystack: Uint8Array,
  needle: Uint8Array,
  start: number = 0
): number {
  if (needle.length === 0) {
    return 0
  }

  // The last possible starting position for a match
  const limit = haystack.length - needle.length

  for (let i = start; i <= limit; i++) {
    let foundMatch = true
    for (let j = 0; j < needle.length; j++) {
      if (haystack[i + j] !== needle[j]) {
        foundMatch = false
        break // Mismatch, break inner loop
      }
    }
    if (foundMatch) {
      return i // Found a full match at index i
    }
  }

  return -1 // No match found
}
