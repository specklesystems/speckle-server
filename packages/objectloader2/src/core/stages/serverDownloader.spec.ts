import { describe, expect, test } from 'vitest'
import createFetchMock from 'vitest-fetch-mock'
import { vi } from 'vitest'
import { Item } from '../../types/types.js'
import ServerDownloader from './serverDownloader.js'
import AsyncGeneratorQueue from '../../queues/asyncGeneratorQueue.js'

describe('downloader', () => {
  test('download batch of one', async () => {
    const fetchMocker = createFetchMock(vi)
    const i: Item = { baseId: 'id', base: { id: 'id', speckle_type: 'type' } }
    fetchMocker.mockResponseOnce('id\t' + JSON.stringify(i.base) + '\n')
    const gathered = new AsyncGeneratorQueue<Item>()
    const downloader = new ServerDownloader({
      serverUrl: 'http://speckle.test',
      streamId: 'streamId',
      objectId: 'objectId',
      token: 'token',
      fetch: fetchMocker,
      logger: (): void => {}
    })
    downloader.initialize({
      results: gathered,
      total: 1,
      maxDownloadBatchWait: 200
    })
    downloader.add('id')
    const r = []
    for await (const x of gathered.consume()) {
      r.push(x)
      if (r.length >= 1) {
        break
      }
    }

    expect(r).toMatchSnapshot()
    await downloader.disposeAsync()
  })

  test('download batch of two', async () => {
    const fetchMocker = createFetchMock(vi)
    const i1: Item = { baseId: 'id1', base: { id: 'id1', speckle_type: 'type' } }
    const i2: Item = { baseId: 'id2', base: { id: 'id2', speckle_type: 'type' } }
    fetchMocker.mockResponseOnce(
      'id1\t' + JSON.stringify(i1.base) + '\nid2\t' + JSON.stringify(i2.base) + '\n'
    )

    const gathered = new AsyncGeneratorQueue<Item>()
    const downloader = new ServerDownloader({
      serverUrl: 'http://speckle.test',
      streamId: 'streamId',
      objectId: 'objectId',
      token: 'token',

      fetch: fetchMocker,
      logger: (): void => {}
    })
    downloader.initialize({
      results: gathered,
      total: 2,
      maxDownloadBatchWait: 200
    })
    downloader.add('id1')
    downloader.add('id2')
    await downloader.disposeAsync()
    const r = []
    for await (const x of gathered.consume()) {
      r.push(x)
      if (r.length >= 2) {
        break
      }
    }

    expect(r).toMatchSnapshot()
    await downloader.disposeAsync()
  })

  test('download batch of three', async () => {
    const fetchMocker = createFetchMock(vi)
    const i1: Item = { baseId: 'id1', base: { id: 'id1', speckle_type: 'type' } }
    const i2: Item = { baseId: 'id2', base: { id: 'id2', speckle_type: 'type' } }
    const i3: Item = { baseId: 'id3', base: { id: 'id3', speckle_type: 'type' } }
    fetchMocker.mockResponseOnce(
      'id1\t' +
        JSON.stringify(i1.base) +
        '\nid2\t' +
        JSON.stringify(i2.base) +
        '\nid3\t' +
        JSON.stringify(i3.base) +
        '\n'
    )

    const gathered = new AsyncGeneratorQueue<Item>()
    const downloader = new ServerDownloader({
      serverUrl: 'http://speckle.test',
      streamId: 'streamId',
      objectId: 'objectId',
      token: 'token',

      fetch: fetchMocker,
      logger: (): void => {}
    })
    downloader.initialize({
      results: gathered,
      total: 3,
      maxDownloadBatchWait: 200
    })
    downloader.add('id1')
    downloader.add('id2')
    downloader.add('id3')
    await downloader.disposeAsync()
    const r = []
    for await (const x of gathered.consume()) {
      r.push(x)
      if (r.length >= 3) {
        break
      }
    }

    expect(r).toMatchSnapshot()
    await downloader.disposeAsync()
  })

  test('download batch of three with Objects.Other.RawEncoding', async () => {
    const fetchMocker = createFetchMock(vi)
    const i1: Item = { baseId: 'id1', base: { id: 'id1', speckle_type: 'type' } }
    const i2: Item = {
      baseId: 'id2',
      base: { id: 'id2', speckle_type: 'Objects.Other.RawEncoding' }
    }
    const i3: Item = { baseId: 'id3', base: { id: 'id3', speckle_type: 'type' } }
    fetchMocker.mockResponseOnce(
      'id1\t' +
        JSON.stringify(i1.base) +
        '\nid2\t' +
        JSON.stringify(i2.base) +
        '\nid3\t' +
        JSON.stringify(i3.base) +
        '\n'
    )

    const gathered = new AsyncGeneratorQueue<Item>()
    const downloader = new ServerDownloader({
      serverUrl: 'http://speckle.test',
      streamId: 'streamId',
      objectId: 'objectId',
      token: 'token',

      fetch: fetchMocker,
      logger: (): void => {}
    })
    downloader.initialize({
      results: gathered,
      total: 3,
      maxDownloadBatchWait: 200
    })
    downloader.add('id1')
    downloader.add('id2')
    downloader.add('id3')
    await downloader.disposeAsync()
    const r = []
    for await (const x of gathered.consume()) {
      r.push(x)
      if (r.length >= 3) {
        break
      }
    }

    expect(r).toMatchSnapshot()
    await downloader.disposeAsync()
  })

  test("download Objects.Other.RawEncoding doesn't exist", async () => {
    const fetchMocker = createFetchMock(vi)
    const i: Item = {
      baseId: 'id',
      base: {
        id: 'id',
        speckle_type: 'Objects.Other.RawEncoding',
        __closure: { childIds: 1 }
      }
    }
    fetchMocker.mockResponseOnce(JSON.stringify(i.base))
    const downloader = new ServerDownloader({
      serverUrl: 'http://speckle.test',
      streamId: 'streamId',
      objectId: i.baseId,
      token: 'token',

      fetch: fetchMocker,
      logger: (): void => {}
    })
    await expect(downloader.downloadSingle()).rejects.toThrow()
    await downloader.disposeAsync()
  })

  test('download single exists', async () => {
    const fetchMocker = createFetchMock(vi)
    const i: Item = {
      baseId: 'id',
      base: { id: 'id', speckle_type: 'type', __closure: { childIds: 1 } }
    }
    fetchMocker.mockResponseOnce(JSON.stringify(i.base))
    const downloader = new ServerDownloader({
      serverUrl: 'http://speckle.test',
      streamId: 'streamId',
      objectId: i.baseId,
      token: 'token',

      fetch: fetchMocker,
      logger: (): void => {}
    })
    const x = await downloader.downloadSingle()
    expect(x).toMatchSnapshot()
    await downloader.disposeAsync()
  })

  test('add extra header', async () => {
    const fetchMocker = createFetchMock(vi)
    const i: Item = {
      baseId: 'id',
      base: { id: 'id', speckle_type: 'type', __closure: { childIds: 1 } }
    }
    fetchMocker.mockResponseIf(
      (req) => req.headers.get('x-test') === 'asdf',
      JSON.stringify(i.base)
    )
    const headers = new Headers()
    headers.set('x-test', 'asdf')
    const downloader = new ServerDownloader({
      serverUrl: 'http://speckle.test',
      headers,
      streamId: 'streamId',
      objectId: i.baseId,
      token: 'token',

      fetch: fetchMocker,
      logger: (): void => {}
    })
    const x = await downloader.downloadSingle()
    expect(x).toMatchSnapshot()
    await downloader.disposeAsync()
  })

  test('can dispose used', async () => {
    const fetchMocker = createFetchMock(vi)
    const headers = new Headers()
    const downloader = new ServerDownloader({
      serverUrl: 'http://speckle.test',
      headers,
      streamId: 'streamId',
      objectId: 'objectId',
      token: 'token',

      fetch: fetchMocker,
      logger: (): void => {}
    })
    await downloader.disposeAsync()
  })

  test('nothing is frozen when validateResponse returns 403', async () => {
    const fetchMocker = createFetchMock(vi)
    const consoleErrorSpy = vi.spyOn(console, 'error').mockImplementation(() => {})

    // Mock a 403 Forbidden response
    fetchMocker.mockResponseOnce('', { status: 403, statusText: 'Forbidden' })

    const gathered = new AsyncGeneratorQueue<Item>()
    const downloader = new ServerDownloader({
      serverUrl: 'http://speckle.test',
      streamId: 'streamId',
      objectId: 'objectId',
      token: 'invalid-token',
      fetch: fetchMocker,
      logger: (): void => {}
    })

    try {
      downloader.initialize({
        results: gathered,
        total: 2,
        maxDownloadBatchWait: 100
      })

      // Add items to trigger batch processing
      downloader.add('id1')
      downloader.add('id2')

      // Wait for the batch to be processed and fail with 403
      await new Promise((resolve) => setTimeout(resolve, 200))

      // Verify that the error was logged (indicating the batch processing failed)
      expect(consoleErrorSpy).toHaveBeenCalledWith(
        'Batch processing failed:',
        expect.any(Error)
      )

      // The key test: verify we can still dispose the downloader properly
      // This ensures the system isn't frozen and can clean up resources
      const disposePromise = downloader.disposeAsync()

      // Add a timeout to ensure disposal doesn't hang indefinitely
      const timeoutPromise = new Promise((_, reject) => {
        setTimeout(() => reject(new Error('Disposal timed out')), 5000)
      })

      // This should complete without timing out or throwing
      await Promise.race([disposePromise, timeoutPromise])

      // Additional verification: the batching queue should be marked as disposed
      // We can't directly access the private field, but we can verify disposal completed
      expect(true).toBe(true) // If we reach here, disposal succeeded
    } finally {
      consoleErrorSpy.mockRestore()
    }
  })

  test('system remains functional after 403 error and can be properly cleaned up', async () => {
    const fetchMocker = createFetchMock(vi)
    const consoleErrorSpy = vi.spyOn(console, 'error').mockImplementation(() => {})

    // First call returns 403, subsequent calls should not be made due to queue disposal
    fetchMocker.mockResponseOnce('', { status: 403, statusText: 'Forbidden' })

    const gathered = new AsyncGeneratorQueue<Item>()
    const downloader = new ServerDownloader({
      serverUrl: 'http://speckle.test',
      streamId: 'streamId',
      objectId: 'objectId',
      token: 'invalid-token',
      fetch: fetchMocker,
      logger: (): void => {}
    })

    try {
      downloader.initialize({
        results: gathered,
        total: 5,
        maxDownloadBatchWait: 50
      })

      // Add first batch that will trigger the 403 error
      downloader.add('id1')
      downloader.add('id2')

      // Wait for first batch to fail
      await new Promise((resolve) => setTimeout(resolve, 100))

      // Verify error was logged
      expect(consoleErrorSpy).toHaveBeenCalledWith(
        'Batch processing failed:',
        expect.any(Error)
      )

      // Try to add more items after the failure
      // These should be ignored since the queue is now disposed
      downloader.add('id3')
      downloader.add('id4')
      downloader.add('id5')

      // Wait a bit more to ensure no additional processing attempts
      await new Promise((resolve) => setTimeout(resolve, 100))

      // Note: The batching queue might make multiple attempts before disposal
      // The key is that disposal should still work regardless of how many calls were made
      expect(fetchMocker).toHaveBeenCalled()

      // Critical test: disposal should complete without hanging
      const start = Date.now()
      await downloader.disposeAsync()
      const elapsed = Date.now() - start

      // Disposal should be quick (under 1 second) and not hang
      expect(elapsed).toBeLessThan(1000)

      // Verify that the results queue can also be disposed properly
      await gathered.disposeAsync()
    } finally {
      consoleErrorSpy.mockRestore()
    }
  })
})
