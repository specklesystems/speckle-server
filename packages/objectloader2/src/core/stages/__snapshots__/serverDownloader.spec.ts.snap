// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`downloader > add extra header 1`] = `
{
  "base": {
    "__closure": {
      "childIds": 1,
    },
    "id": "id",
    "speckle_type": "type",
  },
  "baseId": "id",
  "size": 0,
}
`;

exports[`downloader > download batch of one 1`] = `
[
  {
    "base": {
      "id": "id",
      "speckle_type": "type",
    },
    "baseId": "id",
    "size": 33,
  },
]
`;

exports[`downloader > download batch of three 1`] = `
[
  {
    "base": {
      "id": "id1",
      "speckle_type": "type",
    },
    "baseId": "id1",
    "size": 34,
  },
  {
    "base": {
      "id": "id2",
      "speckle_type": "type",
    },
    "baseId": "id2",
    "size": 34,
  },
  {
    "base": {
      "id": "id3",
      "speckle_type": "type",
    },
    "baseId": "id3",
    "size": 34,
  },
]
`;

exports[`downloader > download batch of two 1`] = `
[
  {
    "base": {
      "id": "id1",
      "speckle_type": "type",
    },
    "baseId": "id1",
    "size": 34,
  },
  {
    "base": {
      "id": "id2",
      "speckle_type": "type",
    },
    "baseId": "id2",
    "size": 34,
  },
]
`;

exports[`downloader > download single exists 1`] = `
{
  "base": {
    "__closure": {
      "childIds": 1,
    },
    "id": "id",
    "speckle_type": "type",
  },
  "baseId": "id",
  "size": 0,
}
`;
